src/coreComponents/constitutive/solid/InvariantDecompositions.hpp

```diff
@@ -0,0 +1,222 @@
+ *
+ * Stress and strain invariants are typically used in constitutive modeling
+ * to define a frame-indifferent response.
+ *
+ * @note
+ * Care must be used in interpreting the decompositions, as Voight
+ * notation uses a different convention for stress vs strain.  In particular,
+ * the deviator (\f$\mathbf{n}\f$) is stored as a
+ * stress-like quantity, such that it satisfies the relationship
+ *        \f[
+ *        \sigma = p \mathbf{I} + \sqrt{2/3} q \mathbf{n}
+ *        \f]
+ * when written in "unrolled" form.  It does not immediately satisfy
+ *        \f[
+ *        \epsilon = \frac{1}{3} \epsilon_v \mathbf{I} + \sqrt{3/2} \epsilon_d \mathbf{n}
+ *        \f]
+ * due to the engineering notation used for off-diagonal strains.  The stress and
+ * strain recomposition functions account for this subtlety, so the user should
+ * be safe when using these utilities directly.
+ */
+namespace twoInvariant
+{
+
+/**
+ * @brief Perform two-invariant decomposition of strain tensors
+ *
+ * @param[in] strain Strain tensor in Voight notation
+ * @param[out] volStrain Volumetric strain invariant
+ * @param[out] devStrain Deviatoric strain invariant
+ * @param[out] deviator Unit orientation (tensor) for deviatoric part in "stress" Voight notation
+ */
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void strainDecomposition( real64 const ( &strain )[6],
+                          real64 & volStrain,
+                          real64 & devStrain,
+                          real64 ( & deviator )[6] )
+{
+  volStrain = 0;
+  for( localIndex i=0; i<3; ++i )
+  {
+    volStrain += strain[i];
+  }
+
+  for( localIndex i=0; i<3; ++i )
+  {
+    deviator[i] = strain[i] - volStrain/3.;
+    deviator[i+3] = strain[i+3]/2.; // divide by two ("stress voight")
+  }
+
+  devStrain = 0;
+  for( localIndex i=0; i<3; ++i )
+  {
+    devStrain += deviator[i] * deviator[i];
+    devStrain += 2 * deviator[i+3] * deviator[i+3];
+  }
+  devStrain = std::sqrt( devStrain );
+
+  if( devStrain < 1e-12 )
+  {
+    for( localIndex i=0; i<6; ++i )
+    {
+      deviator[i] = 0;
+    }
+  }
+  else
+  {
+    for( localIndex i=0; i<6; ++i )
+    {
+      deviator[i] /= devStrain;
+    }
+  }
+  devStrain *= sqrt( 2./3. );
+  return;
+}
+
+
+/**
+ * @brief Perform two-invariant decomposition of stress tensors
+ *
+ * @param[in] stress Stress tensor in Voight notation
+ * @param[out] volStress Volumetric stress invariant (= mean stress)
+ * @param[out] devStress Deviatoric stress invariant (= von Mises stress)
+ * @param[out] deviator Unit orientation (tensor) for deviatoric part in "stress" Voight notation
+ */
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void stressDecomposition( real64 const ( &stress )[6],
+                          real64 & volStress,
+                          real64 & devStress,
+                          real64 ( & deviator )[6] )
+{
+  volStress = 0;
+  for( localIndex i=0; i<3; ++i )
+  {
+    volStress += stress[i];
+  }
+  volStress /= 3.;
+
```

Comment:
Just one line for this ^^ : volStress = ( stress[0] + stress[1] + stress[2] ) / 3.0